include/circt/Dialect/Calyx/CalyxStructure.td

```diff
@@ -197,24 +197,38 @@ def GroupOp : CalyxOp<"group", [
   let assemblyFormat = "$sym_name $body attr-dict";
 }
 
-// TODO(Calyx): Add optional guard to AssignOp and DoneOp.
+class SameTypeConstraint<string lhs, string rhs>
+ : TypesMatchWith<"lhs and rhs types should be equivalent",
+                   lhs, rhs, [{ $_self }]>;
+
 def AssignOp : CalyxOp<"assign", [
-    SameTypeOperands
+    SameTypeConstraint<"dest", "src">
   ]> {
   let summary = "Calyx Assignment";
   let description = [{
     The "calyx.assign" operation represents a non-blocking
-    assignment. An assignment may optionally be
-    guarded. This operation should only be instantiated in the
+    assignment. An assignment may optionally be guarded,
+    which controls when the assignment should be active.
+    This operation should only be instantiated in the
     "calyx.wires" section or a "calyx.group".
 
     ```mlir
       calyx.assign %1 = %2 : i16
+      calyx.assign %1 = %2, %guard ? : i16
     ```
   }];
-  let arguments = (ins AnyType:$dest, AnyType:$src);
-  let assemblyFormat = "$dest `=` $src attr-dict `:` type($src)";
+  let arguments = (ins
+    AnyType:$dest,
+    AnyType:$src,
+    Optional<I1>:$guard
+  );
   let verifier = "return ::verify$cppClass(*this);";
+
+  // TODO(Calyx): Calyx IR typically represents a
+  // guarded assignments as `%dest = %guard ? %src`
+  let assemblyFormat = [{
+    $dest `=` $src (`,` $guard^ `?`)? attr-dict `:` type($dest)
```

Comment:
I had another thought about this, not sure if it works... could the syntax here be:

```suggestion
    $dest `=` $src (`?` $guard^)? attr-dict `:` type($dest)
```

Not sure if the MLIR parser generator will accept that, but it could save some punctuation, and the form would be similar to the existing Calyx form, but with the optional guard location flipped. I.e.

```
existing: dest = guard ? src
new: dest = src ? guard
```

Reply:
The MLIR parser generator should accept this, but I'm personally not a fan, given the imminent confusion with the existing IR.

CC @rachitnigam, @sampsyo to see if they have a strong opinion in either direction.